### PR TITLE
fix(require-lookup): Use a safer way of extracting exports

### LINKED ISSUE #126: Atom crashes with autocomplete-modules if a `require()`d file contains `process.exit()`
### Prerequisites

* [x] Put an X between the brackets on this line if you have done all of the following:
    * [ ] Reproduced the problem in Safe Mode: http://flight-manual.atom.io/hacking-atom/sections/debugging/#using-safe-mode
    * [x] Followed all applicable steps in the debugging guide: http://flight-manual.atom.io/hacking-atom/sections/debugging/
    * [x] Checked the FAQs on the message board for common solutions: https://discuss.atom.io/c/faq
    * [x] Checked that your issue isn't already filed: https://github.com/issues?utf8=✓&q=is%3Aissue+user%3Aatom
    * [x] Checked that there is not already an Atom package that provides the described functionality: https://atom.io/packages

### Description

I'm writing a command-line tool that has a module that checks for a minimum Node version and exits if the version doesn't match.

`ensurenodeversion.js`
```js
if (!/^1[012]\./.test(process.versions.node)) {
  process.stderr.write(
    'Expected Node version at least 10.0.0, got ' + process.versions.node + '\n');
  process.exit(-1); // eslint-disable-line no-process-exit
}
```

Suddenly, Atom started crashing reproducibly when I typed a comma after a destructuring from a `require` return value.

`extractpdfinfo.js`
```js
require('./ensurenodeversion');

const argv = require('minimist')(process.argv.slice(2));

const { handleError } = require('./utils');
const { extractPdfDocumentInfo, analyzePdfDocument, } = require('./reportfiles');
//                                                 \
//                                                  \_____ THIS COMMA
```

I added a stack trace to where `process.exit()` was called:
```js
if (!/^1[012]\./.test(process.versions.node)) {
  process.stderr.write(
    'Expected Node version at least 10.0.0, got ' +
      process.versions.node +
      '\n' +
      require('util').inspect(process.argv) +
      ' ' +
      new Error().stack,
  );
  //process.exit(-1); // eslint-disable-line no-process-exit
}
```

and got this:
```
Expected Node version at least 10.0.0, got 8.9.3
[ '/Applications/Atom.app/Contents/Frameworks/Atom Helper.app/Contents/MacOS/Atom Helper',
  '--type=renderer',
  '--enable-experimental-web-platform-features',
  '--no-sandbox',
  '--service-pipe-token=4E5A7935C4CCC03721B91B013C41CD51',
  '--lang=en-US',
  '--app-path=/Applications/Atom.app/Contents/Resources/app.asar',
  '--node-integration=true',
  '--webview-tag=true',
  '--no-sandbox',
  '--disable-blink-features=Auxclick',
  '--context-id=2',
  '--enable-pinch',
  '--num-raster-threads=4',
  '--enable-zero-copy',
  '--enable-gpu-memory-buffer-compositor-resources',
  '--enable-main-frame-before-activation',
  '--content-image-texture-target=0,0,3553;0,1,3553;0,2,3553;0,3,3553;0,4,3553;0,5,3553;0,6,3553;0,7,3553;0,8,3553;0,9,3553;0,10,3553;0,11,34037;0,12,34037;0,13,34037;0,14,3553;0,15,3553;0,16,3553;0,17,3553;1,0,3553;1,1,3553;1,2,3553;1,3,3553;1,4,3553;1,5,3553;1,6,3553;1,7,3553;1,8,3553;1,9,3553;1,10,3553;1,11,34037;1,12,34037;1,13,34037;1,14,3553;1,15,3553;1,16,3553;1,17,3553;2,0,3553;2,1,3553;2,2,3553;2,3,3553;2,4,3553;2,5,3553;2,6,3553;2,7,3553;2,8,3553;2,9,3553;2,10,3553;2,11,34037;2,12,34037;2,13,34037;2,14,3553;2,15,3553;2,16,3553;2,17,3553;3,0,3553;3,1,3553;3,2,3553;3,3,3553;3,4,3553;3,5,34037;3,6,3553;3,7,3553;3,8,3553;3,9,3553;3,10,3553;3,11,3553;3,12,3553;3,13,34037;3,14,34037;3,15,3553;3,16,34037;3,17,34037;4,0,3553;4,1,3553;4,2,3553;4,3,3553;4,4,3553;4,5,34037;4,6,3553;4,7,3553;4,8,3553;4,9,3553;4,10,3553;4,11,3553;4,12,3553;4,13,34037;4,14,34037;4,15,3553;4,16,34037;4,17,34037',
  '--enable-gpu-async-worker-context',
  '--service-request-channel-token=4E5A7935C4CCC03721B91B013C41CD51',
  '--renderer-client-id=4' ] Error
    at Object.<anonymous> (/__CENSORED__/src/ensurenodeversion.js:8:7)
    at Object.<anonymous> (/__CENSORED__/src/ensurenodeversion.js:13:3)
    at Module._compile (/Applications/Atom.app/Contents/Resources/app.asar/src/native-compile-cache.js:87:30)
    at Object.value [as .js] (/Applications/Atom.app/Contents/Resources/app.asar/src/compile-cache.js:235:23)
    at Module.load (module.js:561:32)
    at tryModuleLoad (module.js:504:12)
    at Function.Module._load (module.js:496:3)
    at Module.require (module.js:586:17)
    at require (/Applications/Atom.app/Contents/Resources/app.asar/src/native-compile-cache.js:47:27)
    at Object.<anonymous> (/__CENSORED__/src/reportfiles.js:1:156)
    at Object.<anonymous> (/__CENSORED__/src/reportfiles.js:1077:3)
    at Module._compile (/Applications/Atom.app/Contents/Resources/app.asar/src/native-compile-cache.js:87:30)
    at Object.value [as .js] (/Applications/Atom.app/Contents/Resources/app.asar/src/compile-cache.js:235:23)
    at Module.load (module.js:561:32)
    at tryModuleLoad (module.js:504:12)
    at Function.Module._load (module.js:496:3)
    at Module.require (module.js:586:17)
    at require (/Applications/Atom.app/Contents/Resources/app.asar/src/native-compile-cache.js:47:27)
    at LocalLookup.lookupCommonjs (/Users/ivanbabak/.atom/packages/autocomplete-modules/lib/utils/lookup-commonjs.js:6:25)
    at getImportModuleAndExt.then.then.results (/Users/ivanbabak/.atom/packages/autocomplete-modules/lib/lookups/export/local.js:25:16)
    at tryCatcher (/Users/ivanbabak/.atom/packages/autocomplete-modules/node_modules/bluebird/js/release/util.js:16:23)
    at Promise._settlePromiseFromHandler (/Users/ivanbabak/.atom/packages/autocomplete-modules/node_modules/bluebird/js/release/promise.js:512:31)
    at Promise._settlePromise (/Users/ivanbabak/.atom/packages/autocomplete-modules/node_modules/bluebird/js/release/promise.js:569:18)
    at Promise._settlePromise0 (/Users/ivanbabak/.atom/packages/autocomplete-modules/node_modules/bluebird/js/release/promise.js:614:10)
    at Promise._settlePromises (/Users/ivanbabak/.atom/packages/autocomplete-modules/node_modules/bluebird/js/release/promise.js:693:18)
    at Async._drainQueue (/Users/ivanbabak/.atom/packages/autocomplete-modules/node_modules/bluebird/js/release/async.js:133:16)
    at Async._drainQueues (/Users/ivanbabak/.atom/packages/autocomplete-modules/node_modules/bluebird/js/release/async.js:143:10)
    at Async.drainQueues (/Users/ivanbabak/.atom/packages/autocomplete-modules/node_modules/bluebird/js/release/async.js:17:14)
    at <anonymous>
```

So I figured when I typed a comma, this triggered `autocomplete-modules`, which in turn **runs my files** to determine what they export.

On top of that, as my other file (`reportfiles.js`) installs a `SIGINT` handler, this gets installed **into Atom**, so if I launch Atom from terminal (`atom -d -f`), I get this in the output (which is output of my SIGINT handler):
```
^C


Got SIGINT, will terminate gracefully. Please wait.



```

### Steps to Reproduce

See above.

#### Folder structure and import statement

```
.                    // Root
├── src
│   ├── ensurenodeversion.js
│   ├── utils.js
│   ├── reportfiles.js
│   └── extractpdfinfo.js   // entry point
```

**Expected behavior:**

I absolutely expect the code editor to **not run my files without my consent** while editing or saving, especially **in the context of Atom editor** (without any isolation), because they may contain **side effects**. Only static analysis unless a run is explicitly requested with a command!

**Actual behavior:** 

`autocomplete-modules` runs the files assuming they have no side effects. Atom crash is just an indicator of this behavior.

**Reproduces how often:**

Every time I enter the comma given the above files.

#### Console errors

See above.

### Versions

```
apm ls | grep autocomplete-modules
├── autocomplete-modules@2.2.6
```

```
atom --version
Atom    : 1.33.0
Electron: 2.0.11
Chrome  : 61.0.3163.100
Node    : 8.9.3
```

```
apm --version
apm  2.1.2
npm  6.2.0
node 8.9.3 x64
atom 1.33.0
python 2.7.15
git 2.14.3
```

OS: `macOS 10.12 Sierra`

### Additional Information

N/A

### PR DESCRIPTION
**Pull request by jonyeezs**
Requiring a file was also executing files. that behaviour is a security issue

Resolves #126 & #100 issue on `require` in code

### CODE REVIEW

File: lib/utils/lookup-commonjs.js [hunk 1]
```diff
@@ -1,12 +1,54 @@
...
+    return getExports(ast);
   } catch(_e) {
       return [];
   }
+
+  function getExports(ast) {
+    return ast.body.reduce((exports, node) => {
+      if(node.type === 'ExpressionStatement') {
+         if (node.expression.left.object.name === 'module' &&
+            node.expression.left.property.name === 'exports')
```
[the remark this answers is not in the document]
Comment (jonyeezs) [reply lib/utils/lookup-commonjs.js]: @sompylasar ah yes! 
I'll make sure to accomadate https://nodejs.org/api/modules.html

File: lib/utils/path-helpers.js [hunk 2]
```diff
@@ -17,6 +17,10 @@ const pathHelpers = {
     return moduleName.endsWith('/') ? ''
       : moduleName.replace(path.dirname(moduleName), '').replace('/', '');
   },
+  getFileDir: (filePath) => {
+    const PATH_SLASH = process.platform === 'win32' ? '\\' : '/';
+    return filePath.substring(0, filePath.lastIndexOf(PATH_SLASH));
```
Comment (sompylasar) [inline lib/utils/path-helpers.js]: What's the difference between this custom impl and `path.dirname`?

Comment (jonyeezs) [reply lib/utils/path-helpers.js]: good point. Not sure of a reasoning behind that.

Was a code carried from the original setup. Might have to do with windows vs linux? I'll look into that later for future refactoring.

**Comment by jonyeezs** [comment]
Can't seem to find a library that would do this for me...sigh might need to write one.

File: lib/utils/path-helpers.js (hunk 2 above)
Comment (sompylasar) [reply lib/utils/path-helpers.js]: As far as I remember, the built-in 'path' module irons out the differences in path representations across platforms, that's why I asked.

**Pull request merged**